src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Utf16StringMarshaller.cs

```diff
@@ -0,0 +1,122 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Runtime.CompilerServices;
+
+namespace System.Runtime.InteropServices
+{
+    /// <summary>
+    /// Marshaller for UTF-16 strings
+    /// </summary>
+    [CLSCompliant(false)]
+    [CustomTypeMarshaller(typeof(string), BufferSize = 0x100,
+        Features = CustomTypeMarshallerFeatures.UnmanagedResources | CustomTypeMarshallerFeatures.TwoStageMarshalling | CustomTypeMarshallerFeatures.CallerAllocatedBuffer)]
+    public unsafe ref struct Utf16StringMarshaller
+    {
+        private ushort* allocated;
+        private Span<ushort> span;
+        private bool isNullString;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Utf16StringMarshaller"/>.
+        /// </summary>
+        /// <param name="str">The string to marshal.</param>
+        public Utf16StringMarshaller(string str)
+            : this(str, default(Span<ushort>))
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Utf16StringMarshaller"/>.
+        /// </summary>
+        /// <param name="str">The string to marshal.</param>
+        /// <param name="buffer">Buffer that may be used for marshalling.</param>
```

Comment:
Same comment about it being critical the span is to pinned or non-managed heap memory.